File: launcher/api/logs.py
"""Tail the server's own log file for the Logs panel.

The handler exposes two read modes via the same endpoint:

- Full tail (no ``after``): the last ``lines`` complete lines, for the
  initial view.
- Incremental (``after``=<byte offset>): only lines appended since that
  offset, so the UI can live-tail cheaply every couple of seconds.

A trailing line that has not been terminated with a newline yet is
withheld and ``next_offset`` only advances past complete lines, so no
partial or duplicated lines ever reach the client. When the file has
shrunken (rotated or truncated) the offset is no longer valid and the
handler falls back to a full tail with ``reset: true``.
"""

from ..config import log_file

DEFAULT_LINES = 500
MAX_LINES = 5000


def handle_list(lines=DEFAULT_LINES, after=None):
    path = log_file()
    try:
        size = path.stat().st_size
    except OSError:
        return {"entries": [], "next_offset": 0, "reset": True}

    try:
        lines = int(lines)
    except (TypeError, ValueError):
        lines = DEFAULT_LINES
    lines = max(1, min(lines, MAX_LINES))

    offset = None
    if after is not None:
        try:
            offset = int(after)
        except (TypeError, ValueError):
            offset = None

    if offset is not None and 0 <= offset <= size:
        return _read_from(path, size, offset)

    return _read_tail(path, size, lines)


def _read_tail(path, size, lines):
    data = _read_bytes(path, 0, size)
    complete, next_offset = _split_complete(data, 0)
    return {"entries": complete[-lines:], "next_offset": next_offset, "reset": True}


def _read_from(path, size, offset):
    data = _read_bytes(path, offset, size - offset)
    complete, next_offset = _split_complete(data, offset)
    return {"entries": complete[-MAX_LINES:], "next_offset": next_offset, "reset": False}


def _read_bytes(path, start, length):
    try:
        with path.open("rb") as f:
            if start:
                f.seek(start)
            return f.read(length) if length > 0 else b""
    except OSError:
        return b""


def _split_complete(data, base):
    """Split a byte chunk into complete lines.

    Returns the decoded lines and the absolute byte offset just past the
    last complete newline. A trailing partial line is withheld.
    """
    last_nl = data.rfind(b"\n")
    if last_nl == -1:
        return [], base
    text = data[: last_nl + 1].decode("utf-8", errors="replace")
    lines = [line[:-1] if line.endswith("\r") else line for line in text.split("\n")[:-1]]
    return lines, base + last_nl + 1

File: internal/api/logs.go
package api

import (
	"os"
	"strconv"
	"strings"

	"launchcontrol/internal/ordjson"
)

// Log tailing defaults, matching the Python logs module.
const (
	defaultLogLines = 500
	maxLogLines     = 5000
)

// TailLogs serves the Logs panel. Two read modes share the endpoint:
// a full tail (no `after`) returning the last `lines` complete lines,
// and an incremental read from a byte offset returning only lines
// appended since then. A trailing unterminated line is withheld and
// next_offset never advances past complete lines, so no partial or
// duplicated lines reach the client. When the file has shrunk (rotated
// or truncated) the handler falls back to a full tail with reset:true.
func (a *API) TailLogs(linesRaw string, afterRaw *string) *ordjson.OMap {
	path := a.LogPath
	info, err := os.Stat(path)
	if err != nil {
		return ordjson.New().Set("entries", []any{}).Set("next_offset", 0).Set("reset", true)
	}
	size := info.Size()

	lines := defaultLogLines
	if n, err := strconv.Atoi(linesRaw); err == nil {
		lines = n
	}
	if lines < 1 {
		lines = 1
	}
	if lines > maxLogLines {
		lines = maxLogLines
	}

	if afterRaw != nil {
		if offset, err := strconv.ParseInt(*afterRaw, 10, 64); err == nil && offset >= 0 && offset <= size {
			return a.readLogsFrom(path, size, offset)
		}
	}
	return a.readLogsTail(path, size, lines)
}

func (a *API) readLogsTail(path string, size int64, lines int) *ordjson.OMap {
	data := readLogBytes(path, 0, size)
	completed, nextOffset := splitCompleteLines(data, 0)
	start := 0
	if len(completed) > lines {
		start = len(completed) - lines
	}
	return ordjson.New().
		Set("entries", toStringsAnyStrings(completed[start:])).
		Set("next_offset", nextOffset).
		Set("reset", true)
}

func (a *API) readLogsFrom(path string, size, offset int64) *ordjson.OMap {
	data := readLogBytes(path, offset, size-offset)
	completed, nextOffset := splitCompleteLines(data, offset)
	start := 0
	if len(completed) > maxLogLines {
		start = len(completed) - maxLogLines
	}
	return ordjson.New().
		Set("entries", toStringsAnyStrings(completed[start:])).
		Set("next_offset", nextOffset).
		Set("reset", false)
}

func readLogBytes(path string, start, length int64) []byte {
	if length <= 0 {
		return nil
	}

	f, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer f.Close()
	if start > 0 {
		if _, err := f.Seek(start, 0); err != nil {
			return nil
		}
	}
	buf := make([]byte, length)
	n, _ := f.Read(buf)
	return buf[:n]
}

// splitCompleteLines decodes a byte chunk into complete lines, keeping
// the offset just past the last newline. A trailing partial line is
// withheld; "\r\n" line endings lose the carriage return.
func splitCompleteLines(data []byte, base int64) ([]string, int64) {
	lastNL := lastByte(data, '\n')
	if lastNL < 0 {
		return nil, base
	}

	text := strings.ToValidUTF8(string(data[:lastNL+1]), "\uFFFD")
	lines := strings.Split(text, "\n")
	// Drop the empty piece after the final newline.
	lines = lines[:len(lines)-1]
	out := make([]string, len(lines))
	for i, line := range lines {
		out[i] = strings.TrimSuffix(line, "\r")
	}
	return out, base + int64(lastNL) + 1
}

func lastByte(data []byte, b byte) int {
	for i := len(data) - 1; i >= 0; i-- {
		if data[i] == b {
			return i
		}
	}
	return -1
}

func toStringsAnyStrings(lines []string) []any {
	out := make([]any, len(lines))
	for i, l := range lines {
		out[i] = l
	}
	return out
}
